### PR TITLE
Tests pools - Healthy pricing behaviour

### PR DESCRIPTION
**Pull request by mustermeiszer**
Tests for ensuring prices of tranches tokens are correct (in a healthy state without defaults)

### CODE REVIEW

File: runtime/development/src/lib.rs [hunk 1]
```diff
@@ -836,7 +836,7 @@ impl pallet_pools::Config for Runtime {
 	type TrancheId = TrancheId;
 	type EpochId = u32;
 	type CurrencyId = CurrencyId;
-	type Tokens = OrmlTokens;
+	type Tokens = Tokens;
```
Comment (mustermeiszer) [inline runtime/development/src/lib.rs:839]: This should be tokens. Otherwise, one can not use a native currency in the pools.

**Comment by mustermeiszer** [comment]
I am sure I won't be able to close the other defined test-cases. But I wanted to note them for the future. Is this alright for you? Otherwise, we could move them to issues.

**Comment by offerijns** [comment]
> I am sure I won't be able to close the other defined test-cases. But I wanted to note them for the future. Is this alright for you? Otherwise, we could move them to issues.
I think it's good to already add them here as placeholders, and then we create a single issue to track actually implementing these.

File: runtime/integration-tests/src/pools/token_prices_normal.rs [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+	let manager = env::task_manager(Handle::current());
+	let mut env = {
+		let mut genesis = Storage::default();
+		genesis::default_balances::<Runtime>(&mut genesis);
+		env::test_env_with_centrifuge_storage::<BLOCK_TIME>(&manager, genesis)
+	};
+
+	let mut nft_manager = NftManager::new();
+	let pool_id = 0u64;
+	let loan_amount = 10_000 * DECIMAL_BASE_12;
```
Comment (offerijns) [inline runtime/integration-tests/src/pools/token_prices_normal.rs:55]:
```suggestion
	let loan_value = 10_000 * DECIMAL_BASE_12;
```

File: runtime/integration-tests/src/pools/token_prices_normal.rs [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+	let investment = 5_000 * DECIMAL_BASE_12;
+	let maturity = time::date(90 * SECONDS_PER_DAY);
+
+	env::run!(
+		env,
+		Chain::Para(PARA_ID),
+		Call,
+		ChainState::PoolEmpty,
+		Keyring::Admin => default_pool_calls(Keyring::Admin.into(), pool_id, &mut nft_manager),
+			issue_default_loan(
```
Comment (offerijns) [inline runtime/integration-tests/src/pools/token_prices_normal.rs:68]: I missed this in the other PR, but maybe we should rename this to 
```suggestion
			issue_bullet_loan(
```

Once we have many loan types, it won't be clear what the default one is.

File: runtime/integration-tests/src/pools/token_prices_normal.rs [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+		Event::Pools(pallet_pools::Event::EpochClosed(id, ..)) if [id == pool_id],
+		Event::Loans(pallet_loans::Event::Borrowed(id, loan_id, amount))
+			if [id == pool_id && loan_id == loan_id && amount == borrow_amount],
+	);
+
+	env::pass_n(&mut env, 30 * time::blocks_per_day::<BLOCK_TIME>());
+
+	let token_prices = env
+		.with_state(Chain::Para(PARA_ID), || {
+			pools::with_ext::get_tranche_prices(pool_id)
```
Comment (offerijns) [inline runtime/integration-tests/src/pools/token_prices_normal.rs:140]: I didn't understand originally why the token price had already gone up here, but realized later on that you're calling `update_nav` within `get_tranche_prices`. Maybe overly explicit, but wdyt about calling this `update_nav_and_get_tranche_prices`?

File: runtime/integration-tests/src/pools/utils/loans.rs [hunk 5]
```diff
@@ -243,3 +246,42 @@ pub fn mint_nft_call(class: ClassId, instance: InstanceId, owner: AccountId) ->
...
+
+/// Calculates the final debt for a given
+/// principal amount, rate and period.
+/// Compounding happens every second.
+///
+/// Rate should be in percent. E.g. 15 -> 15%APR
+///
+/// Logic: n = second-per-year
+/// * principal * (1 + r/n)^(end-start)
+pub fn calculate_debt(principal: Balance, rate: u64, start: Moment, end: Moment) -> Balance {
```
Comment (offerijns) [inline runtime/integration-tests/src/pools/utils/loans.rs:262]: I think we should try to keep using the same terminology everywhere. In the `loans` pallet we use this as naming now:

```suggestion
pub fn calculate_expected_repayment_amount(borrowed_amount: Balance, interest_rate_per_sec: u64, origination_date: Moment, maturity_date: Moment) -> Balance {
```

File: pallets/pools/src/tranche.rs [hunk 6]
```diff
@@ -179,7 +179,7 @@ where
 	}
 
 	pub fn accrue(&mut self, now: Moment) -> DispatchResult {
-		let delta = now - self.last_updated_interest;
+		let delta = now.saturating_sub(self.last_updated_interest);
```
Comment (offerijns) [inline pallets/pools/src/tranche.rs:182]: 💯

File: runtime/integration-tests/src/pools/token_prices_normal.rs (hunk 3 above)
Comment (mustermeiszer) [reply runtime/integration-tests/src/pools/token_prices_normal.rs:68]: Good thought! Will do.

File: runtime/integration-tests/src/pools/token_prices_normal.rs (hunk 4 above)
Comment (mustermeiszer) [reply runtime/integration-tests/src/pools/token_prices_normal.rs]: Yes, agree! will do that.

**Comment by mustermeiszer** [comment]
@NunoAlexandre Was there lately not a caching issue? Because the failing integration test runs through just fine on my machine...

**Review by offerijns — APPROVED** [review]

**Pull request closed without merging**